### DISCUSSION TITLE
Electron force mqtts instead of wss

### DISCUSSION THREAD

**Discussion by axi92** [mqttjs/MQTT.js#1854] [Q&A] [answered]
We used mqtts before version 5 of the lib and it worked fine. Now we want to upgrade but it always falls back to ws and tries to open a connection with that.
I tried to force it to use mqtt, the only part where I had success was by modifying the source [isWebWorkerEnv](https://github.com/mqttjs/MQTT.js/blob/b207984a46d8baa7863eacc3a9082babff7c9357/src/lib/is-browser.ts)  to return false so it would connect via mqtt and that worked.

So my question is is there a normal way without forking the lib?

I have  to add the electron webPreferences are:

```
        webPreferences: {
            sandbox: false,
            nodeIntegration: true,
            contextIsolation: false,
            zoomFactor: touchZoomFactor,
        },
```

**Comment by robertsLando** [comment]
In order to fix this we should identify electron env and if so return false in isBrowser. Would you like to submit a PR?

**Reply by axi92** [reply]
One way would be to search in the userAgent for "electron"
Maybe check it in isStandardBrowserEnv() because right now it returns true.
![image](https://github.com/mqttjs/MQTT.js/assets/1394837/e16c1380-e7fb-4a3d-a9ce-8b2aaae6adae)

Edit:
If found this maybe thats the way to check https://github.com/electron/electron/issues/2288#issuecomment-337858978

**Reply by robertsLando** [reply]
I think the best way is to do both this checks: https://github.com/electron/electron/issues/2288#issuecomment-611231970

**Reply by axi92** [reply]
I am confused what the transpiled javascript code does there:
![image](https://github.com/mqttjs/MQTT.js/assets/1394837/c6aa3e58-dd4c-4eeb-b02a-933c200c5457)
This returns a boolean https://github.com/mqttjs/MQTT.js/blob/0c72b5a1401d7f92376b7f4cce4d757dea5193ae/src/lib/is-browser.ts#L15
But here it imports an object https://github.com/mqttjs/MQTT.js/blob/0c72b5a1401d7f92376b7f4cce4d757dea5193ae/src/lib/connect/index.ts#L9

Am I missing something?
I was able to modify the javascript code to give me the correct output for isStandardBrowserEnv() so it is false.
But when it does the magic with:
```
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
```
it is an object as shown in the screenshot.

**Reply by robertsLando** [reply]
> But here it imports an object

This is ESM, by default it exports the `default` export, with require it would be equal to: 

```js
 const IS_BROWSER = require('../is-browser').default 
```

Anyway that's the compiled code of esbuilf, ignore it, that's something you should not work on, you should work on typescript code.

**Reply by axi92** [reply]
I am having a hard time to execute the tests even on the main branch they fail at different stages.
Is this a known issue?
After several tries I was able to get a green result.

Any thoughts?
https://github.com/mqttjs/MQTT.js/compare/main...axi92:MQTT.js:feature/electron-context

I was able to test it in the render and in the main process from electron. But I would be happy if somebody else could confirm that.

**Reply by robertsLando** [reply]
Open the pr please. About tests them are flacky but it's known

**Reply by axi92** [reply, marked as answer]
https://github.com/mqttjs/MQTT.js/pull/1856#issue-2273441766
